File: files/js/src/main/scala/com/swoval/files/PlatformWatcher.scala
package com.swoval.files
import com.swoval.logging.Logger

private[files] object PlatformWatcher {
  def make(
      followLinks: Boolean,
      registerable: RegisterableWatchService,
      directoryRegistry: DirectoryRegistry,
      logger: Logger
  ): PathWatcher[PathWatchers.Event] = {
    val watcher = new NioPathWatcher(directoryRegistry, registerable, logger)
    if (followLinks) new SymlinkFollowingPathWatcher(watcher, directoryRegistry, logger)
    else watcher
  }
  def make(
      followLinks: Boolean,
      directoryRegistry: DirectoryRegistry,
      logger: Logger
  ): PathWatcher[PathWatchers.Event] = {
    make(followLinks, null, directoryRegistry, logger)
  }
}

File: files/jvm/src/main/java/com/swoval/files/PlatformWatcher.java
package com.swoval.files;

import com.swoval.files.PathWatchers.Event;
import com.swoval.logging.Logger;
import java.io.IOException;

class PlatformWatcher {
  static PathWatcher<Event> make(
      final boolean followLinks, final DirectoryRegistry directoryRegistry, final Logger logger)
      throws InterruptedException, IOException {
    return make(followLinks, RegisterableWatchServices.get(), directoryRegistry, logger);
  }

  static PathWatcher<Event> make(
      final boolean followLinks,
      final RegisterableWatchService registerableWatchService,
      final DirectoryRegistry directoryRegistry,
      final Logger logger)
      throws InterruptedException, IOException {
    final PathWatcher<Event> pathWatcher =
        new NioPathWatcher(directoryRegistry, registerableWatchService, logger);
    return followLinks
        ? new SymlinkFollowingPathWatcher(pathWatcher, directoryRegistry, logger)
        : pathWatcher;
  }
}
